src/main/java/net/dries007/tfc/common/blockentities/BlastFurnaceBlockEntity.java

```diff
@@ -0,0 +1,579 @@
+        if (cachedRecipe != null)
+        {
+            // Then iterate through the list of item entities again, and store (counting), all possible items which match our selected recipe
+            // This time we match both primary inputs, catalysts, and also fuels.
+            final List<ItemEntity> foundInputs = new ArrayList<>(), foundCatalysts = new ArrayList<>(), foundFuels = new ArrayList<>();
+            int inputCount = 0, catalystCount = 0, fuelCount = 0;
+            for (ItemEntity entity : items)
+            {
+                // Note here, we check that an item matches the recipe as a primary input first
+                // If a specific item counts as both a primary input and a catalyst somehow, we only end up checking it as a primary input
+                // We check fuels last
+                final ItemStack stack = entity.getItem();
+                if (cachedRecipe.matchesInput(stack))
+                {
+                    foundInputs.add(entity);
+                    inputCount += stack.getCount();
+                }
+                else if (cachedRecipe.matchesCatalyst(stack))
+                {
+                    foundCatalysts.add(entity);
+                    catalystCount += stack.getCount();
+                }
+                else if (Helpers.isItem(stack, TFCTags.Items.BLAST_FURNACE_FUEL))
+                {
+                    foundFuels.add(entity);
+                    fuelCount += stack.getCount();
+                }
+            }
+
+            // Now, based on the number of both inputs and catalysts we have identified, we try and insert as much as we can in pairs, up to the capacity
+            // Note that the number of items and the number of catalyst items will always match, and each stack in the inputStacks + catalystStacks should have stack size = 1
+            final int totalInsertCapacity = IntStream.of(
+                capacity - inputStacks.size(),
+                inputCount,
+                catalystCount
+            ).min().orElse(0);
+
+            Helpers.consumeItemsFromEntitiesIndividually(foundInputs, totalInsertCapacity, inputStacks::add);
+            Helpers.consumeItemsFromEntitiesIndividually(foundCatalysts, totalInsertCapacity, catalystStacks::add);
+
+            // Fuels aren't restricted from being 1-1 with catalysts or inputs, so they get consumed independently, up to the same capacity
+            final int totalFuelCapacity = Math.min(
+                capacity - fuelStacks.size(),
+                fuelCount
+            );
+
+            Helpers.consumeItemsFromEntitiesIndividually(foundFuels, totalFuelCapacity, fuelStacks::add);
+
+            markForSync();
+        }
+    }
+
+    /**
+     * Sets a molten block inside the bloomery structure. If there is nothing in the bloomery, attempts to delete any molten blocks left over.
+     */
+    private void updateMoltenBlock(boolean cooking)
```

Comment:
can this be made common to the bloomery? The bloomery slag layers have had reports of not working quite correctly.

Reply:
This could also be a gametest